### ISSUE TITLE
set_probe_matching_kludge is required for overlayfs

### ISSUE THREAD

**Issue by dalehamel** [iovisor/bcc#2715]
To attach to probes on overlayfs, set_probe_matching_kludge is required:

https://github.com/iovisor/bcc/blob/da59f37f14320c0a20300f2c516336ae9ebe8a30/src/cc/bcc_syms.cc#L619-L623

This is further commented in:

https://github.com/iovisor/bcc/blob/da59f37f14320c0a20300f2c516336ae9ebe8a30/src/cc/api/BPF.h#L308-L319

The default behavior when looking for module info for USDT probes is to prefer to match by MAJ:MIN rather than by inode ID

This kludge flag causes the inode ID instead in order to verify that the requested USDT symbol is in a particular file, in order to attach the uprobe to it, and possibly modify semaphores USDT semaphores. This was introduced by https://github.com/iovisor/bcc/commit/01ee0b61a854e06cb62272ddfbb4eeaf16e9ad7c, in #2324 

This means that there are now two filesystems that need this kludge, btrfs (as documented) and overlayfs (not documented except, I think, by this issue). Overlayfs is particularly problematic, because it is the default storage for the graph driver used by most production deployments of docker and containers, such as by kubernetes.

To support USDT tracing on these platforms, code will require this kludge with the majority of tools will probably not know to set.

Philosophically, this also feels like a bit of a leaky abstraction - this is a detail of the filesystem that the binary being probed resides on, which shouldn't really need to be a concern of the caller.

I think we should either:

- Remove this kludge entirely, making resolution by inode the default behavior. In what cases is the MAJ MIN preferred? Is it needed to avoid a collision in some cases? How likely is an inode collision, if it is possible?
- Reverse the kludge, to make the behavior to use MAJ MIN, but make by inode the default. Ok, this is super confusing behavior for a consumer
- Detect what filesystem the target binary resides on, and handle this kludge automatically, and remove the public API as it is no longer necessary.

@davemarchevsky @yonghong-song @palmtenor what are your thoughts?

**Comment by dalehamel** [comment]
I discovered this while working on #2710

**Comment by dalehamel** [comment]
This will also be an issue for #2714 as they run in docker for portability

**Comment by dalehamel** [comment]
I realize now after doing a bit more thinking that matching by MAJ/MIN is because inodes may collide between different devices.

For this reason, maybe the best approach is to try and determine the filesystem type for a given device ID, and if it is overlayfs or btrfs (and possibly others), fallback to inode-only matching? Or perhaps inode + some path matching to mitigate this possibility of inode collisions?

**Comment by yonghong-song** [comment]
@davemarchevsky @palmtenor could you also comment on this? Maybe provide some examples on how to use `set_probe_matching_kludge()` for overlayfs will be good?

**Comment by dalehamel** [comment]
@yonghong-song my PR #2710 uses this, it is required in order for it to pass when I run the tests locally or on github actions with Docker. This is why it is included in my PR.

**Comment by yonghong-song** [comment]
Maybe the current implementation is good enough. We probably need to add some documentation on how this could be used for btrfs and overlayfs. Better, if in the library, we can detect the file system is btrfs/overlayfs and warn people and provide hints for alternatives/examples, could this be better?

**Comment by dalehamel** [comment]
>  Better, if in the library, we can detect the file system is btrfs/overlayfs and warn people and provide hints for alternatives/examples, could this be better?

I would prefer this approach of automatically detecting if it is feasible, the current approach adds complexity. For instance, what do I do in bpftrace for USDT now? I must enable the Kludge always, or else have bpftrace do this logic to determine what filesystem the probe is on? This feels like a leaking abstraction.

**Comment by dalehamel** [comment]
https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt has a description of the overlayfs device nuance right in the first few paragraphs:

> The overlay filesystem approach is 'hybrid', because the objects that
appear in the filesystem do not always appear to belong to that filesystem.
In many cases, an object accessed in the union will be indistinguishable
from accessing the corresponding object from the original filesystem.
This is most obvious from the 'st_dev' field returned by stat(2).

> While directories will report an st_dev from the overlay-filesystem,
non-directory objects may report an st_dev from the lower filesystem or
upper filesystem that is providing the object.  Similarly st_ino will
only be unique when combined with st_dev, and both of these can change
over the lifetime of a non-directory object.  Many applications and
tools ignore these values and will not be affected.

I think we fall into the class of "tools and applications that *will* be affected"

**Comment by dalehamel** [comment]
On overlayfs, the results I get are:

```
Looking by inode /proc/129261/root/usr/local/bin/ruby /usr/local/bin/ruby 1926452 1926452
MAJ 0 8 MIN 62 3
```

This means that the overlayfs device is being detected as "null" / "Unnamed devices", per https://www.kernel.org/doc/Documentation/admin-guide/devices.txt

**Comment by dalehamel** [comment]
To detect the filesystem type, we can probably do something like this:

- Look for matching Major and Minor in /proc/diskstats. If there is a match, this gives us the block device name. This can then be compared against the options in /proc/fs to find the filesystem folder name (eg ext4, btrfs, etc) where this block device is located. This should work for the majority of cases
- If the device is of type 0 (or perhaps, not of major type 8), then we default to inode-only matching, or inode + subpath, or some other sane fallback.

This should also allow for BCC to handle future filesystem compatibility issues, should they arise or be discovered, and let us remove the kludge from the public API.

**Comment by davemarchevsky** [comment]
Thanks for linking the OverlayFS docs, I was unaware of that approach to dev/inode.

> I think we should either:
> * Remove this kludge entirely, making resolution by inode the default behavior. In what cases is the MAJ MIN preferred? Is it needed to avoid a collision in some cases? How likely is an inode collision, if it is possible?
> * Reverse the kludge, to make the behavior to use MAJ MIN, but make by inode the default. Ok, this is super confusing behavior for a consumer
> * Detect what filesystem the target binary resides on, and handle this kludge automatically, and remove the public API as it is no longer necessary.

All of these seem like reasonable ideas to me. The first two options seem preferable as they avoid FS-specific handling. I personally prefer the 2nd - "Reverse the kludge" - don't think inode collisions are enough of an issue in practice, but am not an expert in this realm so wanted the extra strictness as a default for peace of mind.

One thing I don't feel great about is removing `set_probe_matching_kludge` from public API - at least not yet. The heterogeneous environments `bcc` is used in make me wary of making opinionated decisions on behalf of the user without giving them a way to adjust, e.g. "we don't think inode collisions will happen so inode matching is the only thing we'll do" or "we trust that you're reporting your fs type accurately, if not you're out of luck". This is why 2nd option feels best to me.

**Comment by dalehamel** [comment]
> I personally prefer the 2nd - "Reverse the kludge" - don't think inode collisions are enough of an issue in practice, but am not an expert in this realm so wanted the extra strictness as a default for peace of mind.

If we go this route, perhaps we can also try and be a little smarter about the fs handling? I do think it's pretty unlikely that inode collisions will be a problem, so reversing the kludge does kind of make sense to me. I'd still like to try and figure out some sort of smarter handling to try and make the kludge unecessary. Perhaps we reverse it *and* try and do automatic detection? This is more complex, but I think is the tidiest abstraction.

> The heterogeneous environments bcc is used in make me wary of making opinionated decisions on behalf of the user without giving them a way to adjust

I can see this argument, but I still thing that this shouldn't be a reasonable concern of the caller. Will they need to compile a different program for each filesystem their executable may run on?

Thanks for weighing in @davemarchevsky I'll take a swing at a PR to fix this :)

**Comment by dalehamel** [comment]
oh also @davemarchevsky do you have any notes about what issues you ran into matching on MAJ/MIN for btrfs? was it similar to overlay? (major type 0 or something similar?)

**Comment by dalehamel** [comment]
@davemarchevsky i've started a draft PR to enact this #2723

**Comment by dalehamel** [comment]
Closed by #2723
